Gems/AtomLyIntegration/CommonFeatures/Code/Source/PostProcess/EditorPostFxSystemComponent.cpp

```diff
@@ -95,18 +95,33 @@ namespace AZ
         {
             AZ_UNUSED(catalogFile);
 
-            // automatically register all layer categories assets
-            AZ::Data::AssetCatalogRequestBus::Broadcast(&AZ::Data::AssetCatalogRequestBus::Events::EnumerateAssets,
-                nullptr,
-                [this](const AZ::Data::AssetId assetId, const AZ::Data::AssetInfo& assetInfo) {
+            // automatically register all layer category assets at Editor startup
+
+            AZStd::vector<AZ::Data::AssetId> layerAssetIds;
+
+            // First run through all the assets and gather up the asset IDs for all layer assets
+            AZ::Data::AssetCatalogRequestBus::Broadcast(
+                &AZ::Data::AssetCatalogRequestBus::Events::EnumerateAssets, nullptr,
+                [&layerAssetIds](const AZ::Data::AssetId assetId, const AZ::Data::AssetInfo& assetInfo)
+                {
                     const auto assetType = azrtti_typeid<EditorPostFxLayerCategoriesAsset>();
                     if (assetInfo.m_assetType == assetType)
                     {
-                        m_layerCategoriesAssetMap[assetId] = AZ::Data::AssetManager::Instance().GetAsset(assetId, assetType, AZ::Data::AssetLoadBehavior::PreLoad);
-                        m_layerCategoriesAssetMap[assetId].QueueLoad();
+                        layerAssetIds.emplace_back(assetId);
                     }
                 },
                 nullptr);
+
+            // Next, trigger all the loads.  This is done outside of EnumerateAssets to ensure that we don't have any deadlocks caused by
+            // lock inversion.  If this thread locks AssetCatalogRequestBus mutex with EnumerateAssets, then locks m_assetMutex in
+            // AssetManager::FindOrCreateAsset, it's possible for those locks to get locked in reverse on a loading thread, causing a
+            // deadlock.
+            for (auto& assetId : layerAssetIds)
+            {
+                m_layerCategoriesAssetMap[assetId] = AZ::Data::AssetManager::Instance().GetAsset(
+                    assetId, azrtti_typeid<EditorPostFxLayerCategoriesAsset>(), AZ::Data::AssetLoadBehavior::PreLoad);
+                m_layerCategoriesAssetMap[assetId].QueueLoad();
```

Comment:
A bit weird to set PreLoad and also QueueLoad.  PreLoad is supposed to cause it to blockingly load, stalling the current thread, until tis done.  Queue Load means to load it asynchronously and call a callback when its done.

Whats the expected outcome here?  Blocking, or async?  If async, you should probably set it to NoLoad and then queue the load.  If blocking it should be sufficient to set it to Preload.  If that's broken, there's a nother bug!

Reply:
Hmm looking at that code its changed a lot, looks like adding it as Preload will no longer cause it to load at all, which is weird, so you still need to queue load